### PR TITLE
Add FAQ endpoint with GET and PUT

### PR DESCRIPTION
**Pull request by Christopher-walsh22**
### Jira Ticket:

BRS-337

### Jira Ticket URL:

[337](https://github.com/orgs/bcgov/projects/49/views/1?pane=issue&itemId=47365823)

### Description:

Add endpoint to API to handle the GET and PUT requests for FAQ pages

### CODE REVIEW

**Review by cameronpettit — COMMENTED** [review]
Handful of nits, otherwise works

File: lambda/writeFaq/index.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    logger.debug(event.body);
+    const obj = JSON.parse(event.body);
+
+    if (event.httpMethod === 'PUT') {
+        if (permissionObject.isAdmin) {
+            return await updateItem(obj);
+        }else {
+            logger.info('Unauthorized');
+            throw 'Unauthorized Access.';
+          }
```
[the remark this answers is not in the document]
Comment (Christopher-walsh22) [reply lambda/writeFaq/index.js]: Addressed in: [commit](https://github.com/bcgov/parks-reso-api/pull/342/commits/8b1af67ef29180daf8311e042c71956953b73078)

File: lambda/writeFaq/index.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ReturnValues: 'ALL_NEW',
+    TableName: TABLE_NAME,
+    ConditionExpression: 'attribute_exists(pk) AND attribute_exists(sk)'
+  };
+
+  if ('faq' in obj) {
+    updateParams.UpdateExpression += ' #text = :faq,';
+    updateParams.ExpressionAttributeValues[':faq'] = { S: faq };
+    updateParams.ExpressionAttributeNames = { '#text': 'text' };
+  }
```
[the remark this answers is not in the document]
Comment (Christopher-walsh22) [reply lambda/writeFaq/index.js]: Addressed in: [commit](https://github.com/bcgov/parks-reso-api/pull/342/commits/8b1af67ef29180daf8311e042c71956953b73078)

File: lambda/readFaq/index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    return sendResponse(200, {});
+  }
+
+  let queryObj = {
+    TableName: TABLE_NAME
+  };
+
+  try {
+    
+    const token = await decodeJWT(event);
```
[the remark this answers is not in the document]
Comment (Christopher-walsh22) [reply lambda/readFaq/index.js]: Addressed in: [commit](https://github.com/bcgov/parks-reso-api/pull/342/commits/8b1af67ef29180daf8311e042c71956953b73078)

File: lambda/readFaq/index.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    const permissionObject = resolvePermissions(token);
+    queryObj.ExpressionAttributeValues = {};
+    queryObj.ExpressionAttributeValues[':pk'] = { S: 'faq' };
+    queryObj.ExpressionAttributeValues[':sk'] = { S: 'faq' };
+    queryObj.KeyConditionExpression = 'pk =:pk AND sk =:sk';
+    let faq = await runQuery(queryObj);
+    return sendResponse(200, faq, context);
+
+  } catch (err) {
+    console.log(err);
```
[the remark this answers is not in the document]
Comment (Christopher-walsh22) [reply lambda/readFaq/index.js]: Addressed in: [commit](https://github.com/bcgov/parks-reso-api/pull/342/commits/8b1af67ef29180daf8311e042c71956953b73078)

File: lambda/writeFaq/index.js [hunk 5]
```diff
@@ -48,28 +37,28 @@ exports.handler = async (event, context) => {
...
+      ReturnValues: 'ALL_NEW',
+      TableName: TABLE_NAME,
+      ConditionExpression: 'attribute_exists(pk)'
+    };
+    if (obj?.faq) {
+      updateParams.UpdateExpression += ' #text = :faq,';
+      updateParams.ExpressionAttributeValues[':faq'] = { S: obj.faq };
+    }
+    updateParams.UpdateExpression = updateParams.UpdateExpression.slice(0, -1);
+    logger.debug('Updating FAQ:', updateParams);
```
Comment (cameronpettit) [inline lambda/writeFaq/index.js:59]: I would refactor as per my last comment:
```js
if (obj?.faq) {
   let updateParams = {
   ...
   // create and insert PUT item
   ...
   return sendResponse(200, AWS.DynamoDB.Converter.unmarshall(Attributes), context);
} else {
   throw new Error("FAQ property is missing in the input object");
}
```
It's 1 less `if` statement to achieve the same effect. Either way, on L41 ensure you are using optional chaining `obj?.faq` as checking for `obj.faq` will break in the event of a bad request object (`obj` has no property `faq`).

Comment (cameronpettit) [reply lambda/writeFaq/index.js:59]: Also curious if we will ever need the option to clear the faq:

Incoming request body:
```json
{
   faq: ''
}
```
As `obj.faq` is falsy in this case - probably not something that needs to be addressed right now.

Comment (Christopher-walsh22) [reply lambda/writeFaq/index.js]: Addressed in : [commit](https://github.com/bcgov/parks-reso-api/pull/342/commits/ffbe60fb51c9bf3299ab351f08c3e5f476b42000)

**Review by marklise — APPROVED** [review]

**Comment by LindsayMacfarlane** [comment]
Looks like PR deployed and there are no outstanding development tickets. Marking as done.

**Pull request merged**